### PR TITLE
Extend crypto traits and implementations with setup_and_export one-shots

### PR DESCRIPTION
**Pull request by kkohbrok**
This PR introduces new two new functions to the `OpenMlsCrypto` trait that are required for the implementation of the external init feature of MLS. In particular, this PR introduces two HPKE one-shot functions that establish a context and export a secret from it. The function for use by the sender outputs the encapsulated context in addition to the exported secret. The corresponding receiver function takes the encapsulated context as additional input and outputs only the exported secret.

### CODE REVIEW

File: traits/src/crypto.rs [hunk 1]
```diff
@@ -113,6 +113,50 @@ pub trait OpenMlsCrypto {
...
+    ///
+    /// The encapsulated secret is returned together with the exported secret.
+    /// If the secret key is missing in an authenticated mode, an error is
+    /// returned.
+    fn hpke_setup_sender_and_export(
+        &self,
+        config: HpkeConfig,
+        pk_r: &[u8],
+        info: &[u8],
+        psk: Option<&[u8]>,
```
Comment (franziskuskiefer) [inline traits/src/crypto.rs:131]: No need for `psk` and `psk_id`. They are never used by MLS.

File: traits/src/crypto.rs [hunk 2]
```diff
@@ -113,6 +113,50 @@ pub trait OpenMlsCrypto {
...
+        psk_id: Option<&[u8]>,
+        sk_s: Option<&[u8]>,
```
Comment (franziskuskiefer) [inline traits/src/crypto.rs:133]: No need for the auth mode. It's not used by MLS.

File: traits/src/crypto.rs [hunk 3]
```diff
@@ -113,6 +113,50 @@ pub trait OpenMlsCrypto {
...
+        exporter_context: &[u8],
+        exporter_length: usize,
+    ) -> Result<(Vec<u8>, Vec<u8>), CryptoError>;
```
Comment (franziskuskiefer) [inline traits/src/crypto.rs:136]: The return values could get an alias or something to distinguish which is which.

File: traits/src/crypto.rs [hunk 4]
```diff
@@ -113,6 +113,50 @@ pub trait OpenMlsCrypto {
...
+    /// HPKE single-shot setup of a receiver and immediate export a secret.
+    ///
+    /// For the base and PSK modes this decapsulates `enc` with the secret key
+    /// `sk_r` of the receiver.
+    /// For the Auth and AuthPSK modes this decapsulates and authenticates `enc`
+    /// with the secret key `sk_r` of the receiver and the senders public key `pk_s`.
+    ///
+    /// Returns the exported secret. If the secret key is missing in an
+    /// authenticated mode, an error is returned.
+    fn hpke_setup_receiver_and_export(
```
Comment (franziskuskiefer) [inline traits/src/crypto.rs:147]: Same comments as above for the optional args.

File: evercrypt_backend/src/provider.rs [hunk 5]
```diff
@@ -296,9 +296,56 @@ impl OpenMlsCrypto for EvercryptProvider {
...
+        exporter_length: usize,
+    ) -> Result<ExporterSecret, CryptoError> {
+        let hpke = Hpke::<HpkeEvercrypt>::new(
+            hpke::Mode::Base,
+            kem_mode(config.0),
+            kdf_mode(config.1),
+            aead_mode(config.2),
+        );
+        let context = hpke
+            .setup_receiver(enc, &sk_r.into(), info, None, None, None)
```
Comment (franziskuskiefer) [inline evercrypt_backend/src/provider.rs:338]: Looking at the `&sk_r.into()` I feel like hpke-rs should have a byte slice API as well. 🤔

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
One nit, otherwise this looks good to me.

File: evercrypt_backend/src/provider.rs (hunk 5 above)
Comment (kkohbrok) [reply evercrypt_backend/src/provider.rs:316]: It would be a bit more generic. But it wouldn't make that much of a difference, I guess?

Comment (franziskuskiefer) [reply evercrypt_backend/src/provider.rs:316]: It avoids some allocations. But it's unrelated to this PR really 😄

**Comment by franziskuskiefer** [comment]
I noticed just now that I did the single shot exporter functions in hpke a while back. We could use them as well
https://docs.rs/hpke-rs/0.0.12/hpke_rs/struct.Hpke.html#method.send_export

**Pull request merged**